### PR TITLE
[PIL-740] CYA - NFM No ID flow

### PR DESCRIPTION
**Pull request by jassalrichy**
Branch PIL-739 has been merged into this branch.

### CODE REVIEW

File: app/controllers/rfm/NoIdNameRegistrationController.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import play.api.libs.json.Json
+import play.api.data.Form
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.rfm.NoIdNameRegistrationView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class NoIdNameRegistrationController @Inject() (
```
Comment (alikouhestani) [inline app/controllers/rfm/NoIdNameRegistrationController.scala:37]: The prefix "noID" is never used in any class name as we usually use the journey's prefix i.e. "Rfm". Can you please rename to "RfmNameRegistrationController" to follow the usual naming convention. Same for spec file

File: app/controllers/rfm/NoIdRegisteredAddressController.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import utils.countryOptions.CountryOptions
+import views.html.rfm.NoIdRegisteredAddressView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class NoIdRegisteredAddressController @Inject() (
```
Comment (alikouhestani) [inline app/controllers/rfm/NoIdRegisteredAddressController.scala:37]: Again, could you please rename to "RfmRegisteredAddress"

File: app/controllers/rfm/NoIdRegisteredAddressController.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  val form: Form[NonUKAddress] = formProvider()
+
+  def onPageLoad(mode: Mode): Action[AnyContent] = (rfmIdentify andThen getData andThen requireData) { implicit request =>
+    val rfmAccessEnabled = appConfig.rfmAccessEnabled
+    if (rfmAccessEnabled) {
+      request.userAnswers
+        .get(RfmNoIdNameRegistrationPage)
+        .map { name =>
+          val preparedForm = request.userAnswers.get(RfmNoIdRegisteredAddressPage) match {
```
Comment (alikouhestani) [inline app/controllers/rfm/NoIdRegisteredAddressController.scala:59]: Again, I would personally remove the "NoID" from the name here and just change to "RfmRegisteredAddressPage" to follow the usual pattern used in other journeys

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  private val checkRouteMap: Page => UserAnswers => Call = {
+    case RfmPillar2ReferencePage      => whichCheckYourAnswerPageSecurityQuestions
+    case RfmRegistrationDatePage      => whichCheckYourAnswerPageSecurityQuestions
+    case RfmNoIdNameRegistrationPage  => whichCheckYourAnswerPageNoIdQuestions
+    case RfmNoIdRegisteredAddressPage => whichCheckYourAnswerPageNoIdQuestions
+    case _                            => _ => controllers.rfm.routes.StartPageController.onPageLoad
+  }
+
+  private def whichCheckYourAnswerPageSecurityQuestions(userAnswers: UserAnswers): Call =
```
Comment (alikouhestani) [inline app/navigation/ReplaceFilingMemberNavigator.scala:54]: Whilst something similar might be needed later on to decide the routing for the submit and review or the other CYA page, this is currently redundant and not correct as the following CYA pages don't display any two same fields.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    userAnswers.get(RfmCheckYourAnswersLogicPage) match {
+      case Some(true) => reviewAndSubmitCheckYourAnswers
+      case _          => securityQuestionsCheckYourAnswers
+    }
+
+  private def whichCheckYourAnswerPageNoIdQuestions(userAnswers: UserAnswers): Call =
```
Comment (alikouhestani) [inline app/navigation/ReplaceFilingMemberNavigator.scala:60]: same as above

File: app/pages/RfmNoIdNameRegistrationPage.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+package pages
+
+import play.api.libs.json.JsPath
+
+case object RfmNoIdNameRegistrationPage extends QuestionPage[String] {
+
+  override def path: JsPath = JsPath \ toString
+
+  override def toString: String = "rfmNfmNameRegistration"
```
Comment (alikouhestani) [inline app/pages/RfmNoIdNameRegistrationPage.scala:25]: remove the "Nfm" from the name "rfmNfmNameRegistration"

File: app/pages/RfmNoIdRegisteredAddressPage.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+package pages
+
+import models.NonUKAddress
+import play.api.libs.json.JsPath
+
+case object RfmNoIdRegisteredAddressPage extends QuestionPage[NonUKAddress] {
+
+  override def path: JsPath = JsPath \ toString
+
+  override def toString: String = "rfmNfmRegisteredAddress"
```
Comment (alikouhestani) [inline app/pages/RfmNoIdRegisteredAddressPage.scala:26]: Same as above, remove"Nfm"

File: conf/messages.en [hunk 8]
```diff
@@ -901,6 +873,43 @@ rfm.primaryNameRegistration.hint = For example, ‘Tax team’ or ‘Ashley Smit
...
+rfm.noIdRegisteredAddress.postcode.error.length = The postal code must be 10 characters or less
+rfm.noIdRegisteredAddress.postcode.error.invalid = Enter a valid UK postal code
+rfm.noIdRegisteredAddress.country.error.required = Select a country
+rfm.noIdRegisteredAddress.country.error.length = The country cannot be more than 35 characters
+rfm.noIdRegisteredAddress.messages.error.addressLine1.required= Enter first line of the address
+rfm.noIdRegisteredAddress.messages.error.addressLine1.length = The first line of the address must be 35 characters or less
+rfm.noIdRegisteredAddress.messages.error.addressLine1.invalid = The first line of the address contains invalid characters
+rfm.noIdRegisteredAddress.messages.error.addressLine2.length = The second line of the address must be 35 characters or less
+rfm.noIdRegisteredAddress.messages.error.addressLine2.invalid = The second line of the address contains invalid characters
+rfm.noIdRegisteredAddress.change.hidden =  the registered office address of the nominated filing member
```
Comment (alikouhestani) [inline conf/messages.en:911]: You have two values set against rfm.noIdRegisteredAddress.change.hidden
remove one

File: test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        status(result) mustEqual OK
+        contentAsString(result) must include("Check your answer")
+        contentAsString(result) must include("Name")
+        contentAsString(result) must include("Address")
+      }
+    }
+
+    "redirect to Journey Recovery page when rfm noId question status is not completed" in {
+      val testConfig = Configuration("features.rfmAccessEnabled" -> true)
```
Comment (alikouhestani) [inline test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala:58]: This toggle is already set to true in the local environment so this is redundant

File: test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        val request = FakeRequest(GET, controllers.rfm.routes.NoIdCheckYourAnswersController.onPageLoad(NormalMode).url)
+        val result  = route(application, request).value
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result) mustBe Some(controllers.routes.UnderConstructionController.onPageLoad.url)
+      }
+    }
+
+    "redirect to Under Construction page on form submission" in {
+      val testConfig = Configuration("features.rfmAccessEnabled" -> true)
```
Comment (alikouhestani) [inline test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala:87]: same here

File: test/controllers/rfm/NoIdRegisteredAddressControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+              ("addressLine2", "Drive"),
+              ("addressLine3", "Newcastle"),
+              ("addressLine4", "North east"),
+              ("postalCode", "hhhhhhhhhhhh"),
+              ("countryCode", "GB")
+            )
+
+        val result = route(application, request).value
+
+        status(result) mustEqual BAD_REQUEST
```
Comment (alikouhestani) [inline test/controllers/rfm/NoIdRegisteredAddressControllerSpec.scala:189]: could ensure the content of the view here too

**Review by alikouhestani — CHANGES_REQUESTED** [review]

File: app/controllers/rfm/NoIdNameRegistrationController.scala (hunk 1 above)
Comment (jassalrichy) [reply app/controllers/rfm/NoIdNameRegistrationController.scala:37]: Completed by @aashik-varahan in branch PIL-739

File: app/controllers/rfm/NoIdRegisteredAddressController.scala (hunk 2 above)
Comment (jassalrichy) [reply app/controllers/rfm/NoIdRegisteredAddressController.scala:37]: Completed by @aashik-varahan in branch PIL-739

File: app/controllers/rfm/NoIdRegisteredAddressController.scala (hunk 3 above)
Comment (jassalrichy) [reply app/controllers/rfm/NoIdRegisteredAddressController.scala:59]: Completed by @aashik-varahan in branch PIL-739

File: app/pages/RfmNoIdRegisteredAddressPage.scala (hunk 7 above)
Comment (jassalrichy) [reply app/pages/RfmNoIdRegisteredAddressPage.scala:26]: Completed by @aashik-varahan in branch PIL-739

File: test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala (hunk 10 above)
Comment (jassalrichy) [reply test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala:87]: Done

File: test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala (hunk 9 above)
Comment (jassalrichy) [reply test/controllers/rfm/NoIdCheckYourAnswersControllerSpec.scala:58]: Done

File: app/pages/RfmNoIdNameRegistrationPage.scala (hunk 6 above)
Comment (jassalrichy) [reply app/pages/RfmNoIdNameRegistrationPage.scala:25]: Completed by @aashik-varahan in branch PIL-739

File: conf/messages.en (hunk 8 above)
Comment (jassalrichy) [reply conf/messages.en:911]: Done

File: app/navigation/ReplaceFilingMemberNavigator.scala (hunk 5 above)
Comment (jassalrichy) [reply app/navigation/ReplaceFilingMemberNavigator.scala:60]: For RfmRegisteredAddress & RfmNameRegistration, the final review and submit CYA page can contain the same two fields. See Mural fig. PIL-767 16.3a
Have updated navigation to redirect to UnderContrustion page, until this page is develeoped.

File: app/navigation/ReplaceFilingMemberNavigator.scala (hunk 4 above)
Comment (jassalrichy) [reply app/navigation/ReplaceFilingMemberNavigator.scala:54]: Done.

File: test/controllers/rfm/NoIdRegisteredAddressControllerSpec.scala (hunk 11 above)
Comment (aashik-varahan) [reply test/controllers/rfm/NoIdRegisteredAddressControllerSpec.scala]: @alikouhestani @jassalrichy - Added Content view to the unit test

**Review by alikouhestani — APPROVED** [review]

**Pull request merged**